node/src/reactor/initializer.rs

```diff
@@ -118,55 +120,67 @@ impl Reactor {
     pub fn stopped_successfully(&self) -> bool {
         self.chainspec_loader.stopped_successfully()
     }
-}
-
-impl reactor::Reactor for Reactor {
-    type Event = Event;
-    type Config = WithDir<validator::Config>;
-    type Error = Error;
 
-    fn new(
-        config: Self::Config,
+    #[cfg(test)]
+    pub(crate) fn new_with_chainspec(
+        config: <Self as reactor::Reactor>::Config,
         registry: &Registry,
-        event_queue: EventQueueHandle<Self::Event>,
-        _rng: &mut NodeRng,
-    ) -> Result<(Self, Effects<Self::Event>), Error> {
-        let chainspec = config
-            .value()
-            .node
-            .chainspec_config_path
-            .clone()
-            .load(config.dir())
-            .map_err(|err| Error::ConfigError(err.to_string()))?;
-
-        chainspec.validate_config();
-
+        event_queue: EventQueueHandle<Event>,
+        chainspec: Chainspec,
+    ) -> Result<(Self, Effects<Event>), Error> {
         let effect_builder = EffectBuilder::new(event_queue);
+        let (chainspec_loader, chainspec_effects) =
+            ChainspecLoader::new_with_chainspec(chainspec, effect_builder);
+        Self::new_with_chainspec_loader(config, registry, chainspec_loader, chainspec_effects)
+    }
 
+    fn new_with_chainspec_loader(
+        config: <Self as reactor::Reactor>::Config,
+        registry: &Registry,
+        chainspec_loader: ChainspecLoader,
+        chainspec_effects: Effects<chainspec_loader::Event>,
```

Comment:
Alternatively, you don't pass in the `chainspec_effects` but append the resulting ones on the call-side above.

Reply:
There are two call-sites, which would mean the code would be pretty much duplicated.  I think it's better like this?

Reply:
OK, fine with me.